File: JavaScript/프로그래머스/2/87946. 피로도/피로도.js
function dfs(dungeons, visited, k, cnt = 0) {
    let max_cnt = cnt;
    
    for (let i=0; i<dungeons.length; i++) {
        let st = dungeons[i][0];
        let need = dungeons[i][1];
        
        if (!visited[i] && k >= st) {
            visited[i] = true;
            const result = dfs(dungeons, visited, k - need, cnt + 1);
            visited[i] = false;
            
            max_cnt = Math.max(max_cnt, result);
        }
    }
    return max_cnt;
}

function solution(k, dungeons) {
    const visited = new Array(dungeons.length).fill(false);
    let answer = dfs(dungeons, visited, k);
    return answer;
}

File: Java/프로그래머스/2/87946. 피로도/피로도.java
class Solution {
    int result = 0;

    public int solution(int k, int[][] dungeons) {
        boolean[] visited = new boolean[dungeons.length];
        dfs(0, k, dungeons, visited);
        return result;
    }

    public void dfs(int depth, int fatigue, int[][] dungeons, boolean[] visited) {
        result = Math.max(result, depth);

        for (int i = 0; i < dungeons.length; i++) {
            int need = dungeons[i][0];
            int cost = dungeons[i][1];

            if (!visited[i] && fatigue >= need) {
                visited[i] = true;
                dfs(depth + 1, fatigue - cost, dungeons, visited);
                visited[i] = false;
            }
        }
    }
}
